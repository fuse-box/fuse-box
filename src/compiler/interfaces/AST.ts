import { IASTScope } from '../Visitor/Visitor';

export interface ASTNode {
  raw?: string;
  type: keyof typeof ASTType;

  $assign_pattern?: boolean;
  $fuse_classInitializers?: Array<ASTNode>;
  $fuse_class_declaration_visited?: boolean;
  $fuse_decoratorForClassIdentifier?: string;
  $fuse_decoratorsCaptured?: boolean;
  $fuse_decoratorsProcessed?: boolean;
  $fuse_visited?: boolean;
  $transpiled?: boolean;
  alternate?: ASTNode;
  base?: ASTNode;
  chain?: ASTNode;
  consequent?: ASTNode;
  decorators?: Array<ASTNode>;
  definite?: boolean;
  elementTypes?: Array<ASTNode>;
  elements?: Array<ASTNode>;
  expressions?: Array<ASTNode>;
  initializer?: ASTNode;
  literal?: ASTNode;
  moduleReference?: ASTNode;
  optional?: boolean;
  prefix?: boolean;
  returnType?: ASTNode;
  scope?: IASTScope;
  shorthand?: boolean;
  source?: ASTNode;
  sourceType?: string;
  specifiers?: Array<ASTNode>;
  test?: ASTNode;
  typeAnnotation?: ASTNode;
  typeName?: ASTNode;
  loc?: { end: { column: number; line: number }; start: { column: number; line: number } };

  imported?: ASTNode;
  members?: Array<ASTNode>;
  method?: boolean;

  JSXElement?: ASTNode;

  abstract?: boolean;
  accessibility?: string;
  argument?: ASTNode;
  arguments?: Array<ASTNode>;
  async?: boolean;
  attributes?: Array<ASTNode>;
  body?: ASTNode | Array<ASTNode>;
  callee?: ASTNode;
  children?: Array<ASTNode>;
  closingElement?: ASTNode;
  computed?: boolean;
  declaration?: ASTNode;
  declarations?: Array<ASTNode>;
  declare?: boolean;
  exported?: ASTNode;
  expression?: any;
  generator?: boolean;
  id?: ASTNode;
  implements?: Array<ASTNode>;
  init?: null | ASTNode;
  key?: ASTNode;
  kind?: string;
  left?: ASTNode;
  local?: ASTNode;
  name?: any;
  object?: ASTNode;
  openingElement?: ASTNode;
  operator?: string;
  parameter?: ASTNode;
  params?: Array<ASTNode>;
  properties?: Array<ASTNode>;
  property?: ASTNode;
  quasi?: ASTNode;
  quasis?: Array<ASTNode>;
  right?: ASTNode;
  static?: boolean;
  superClass?: ASTNode;
  tag?: ASTNode;
  tail?: ASTNode;
  value?: any;
}

export const ASTType = {
  EnumDeclaration: 'TSEnumDeclaration',
  EnumMember: 'TSEnumMember',

  ParameterProperty: 'TSParameterProperty',
  StringKeyword: 'TSStringKeyword',
  TypeAnnotation: 'TSTypeAnnotation',

  AbstractClassProperty: 'TSAbstractClassProperty',
  AbstractMethodDefinition: 'TSAbstractMethodDefinition',
  AsExpression: 'TSAsExpression',
  DeclareFunction: 'TSDeclareFunction',
  DeclareKeyword: 'TSDeclareKeyword',
  InterfaceDeclaration: 'TSInterfaceDeclaration',
  NonNullExpression: 'TSNonNullExpression',
  TypeAliasDeclaration: 'TSTypeAliasDeclaration',
  TypeAssertion: 'TSTypeAssertion',

  ArrayExpression: 'ArrayExpression',
  ArrayPattern: 'ArrayPattern',
  ArrowFunctionExpression: 'ArrowFunctionExpression',
  AssignmentExpression: 'AssignmentExpression',
  AssignmentPattern: 'AssignmentPattern',
  AwaitExpression: 'AwaitExpression',
  BigIntLiteral: 'BigIntLiteral',
  BinaryExpression: 'BinaryExpression',
  Block: 'Block',
  BlockStatement: 'BlockStatement',
  BreakStatement: 'BreakStatement',
  CallExpression: 'CallExpression',
  CatchClause: 'CatchClause',
  ClassBody: 'ClassBody',
  ClassDeclaration: 'ClassDeclaration',
  ClassExpression: 'ClassExpression',
  ClassProperty: 'ClassProperty',
  ConditionalExpression: 'ConditionalExpression',
  ContinueStatement: 'ContinueStatement',
  DebuggerStatement: 'DebuggerStatement',
  Decorator: 'Decorator',
  DoWhileStatement: 'DoWhileStatement',
  EmptyStatement: 'EmptyStatement',
  ExportAllDeclaration: 'ExportAllDeclaration',
  ExportDefaultDeclaration: 'ExportDefaultDeclaration',
  ExportNamedDeclaration: 'ExportNamedDeclaration',
  ExportSpecifier: 'ExportSpecifier',
  ExpressionStatement: 'ExpressionStatement',
  ForInStatement: 'ForInStatement',
  ForOfStatement: 'ForOfStatement',
  ForStatement: 'ForStatement',
  FunctionDeclaration: 'FunctionDeclaration',
  FunctionExpression: 'FunctionExpression',
  HTMLClose: 'HTMLClose',
  HTMLOpen: 'HTMLOpen',
  Identifier: 'Identifier',
  IfStatement: 'IfStatement',
  Import: 'Import',
  ImportDeclaration: 'ImportDeclaration',
  ImportDefaultSpecifier: 'ImportDefaultSpecifier',
  ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
  ImportSpecifier: 'ImportSpecifier',
  JSXAttribute: 'JSXAttribute',
  JSXClosingElement: 'JSXClosingElement',
  JSXClosingFragment: 'JSXClosingFragment',
  JSXElement: 'JSXElement',
  JSXEmptyExpression: 'JSXEmptyExpression',
  JSXExpressionContainer: 'JSXExpressionContainer',
  JSXFragment: 'JSXFragment',
  JSXIdentifier: 'JSXIdentifier',
  JSXMemberExpression: 'JSXMemberExpression',
  JSXOpeningElement: 'JSXOpeningElement',
  JSXOpeningFragment: 'JSXOpeningFragment',
  JSXSpreadAttribute: 'JSXSpreadAttribute',
  JSXSpreadChild: 'JSXSpreadChild',
  JSXText: 'JSXText',
  LabeledStatement: 'LabeledStatement',
  Line: 'Line',
  Literal: 'Literal',
  LogicalExpression: 'LogicalExpression',
  MemberExpression: 'MemberExpression',
  MetaProperty: 'MetaProperty',
  MethodDefinition: 'MethodDefinition',
  NewExpression: 'NewExpression',
  ObjectExpression: 'ObjectExpression',
  ObjectPattern: 'ObjectPattern',
  OptionalCallExpression: 'OptionalCallExpression',
  OptionalMemberExpression: 'OptionalMemberExpression',
  Program: 'Program',
  Property: 'Property',
  RestElement: 'RestElement',
  ReturnStatement: 'ReturnStatement',
  SequenceExpression: 'SequenceExpression',
  SpreadElement: 'SpreadElement',
  Super: 'Super',
  SwitchCase: 'SwitchCase',
  SwitchStatement: 'SwitchStatement',
  TaggedTemplateExpression: 'TaggedTemplateExpression',
  TemplateElement: 'TemplateElement',
  TemplateLiteral: 'TemplateLiteral',
  ThisExpression: 'ThisExpression',
  ThrowStatement: 'ThrowStatement',
  TryStatement: 'TryStatement',
  module: 'module',
  script: 'script',

  AbstractKeyword: 'AbstractKeyword',

  AnyKeyword: 'TSAnyKeyword',
  ArrayType: 'TSArrayType',

  AsyncKeyword: 'AsyncKeyword',
  BigIntKeyword: 'BigIntKeyword',
  BooleanKeyword: 'TSBooleanKeyword',
  CallSignatureDeclaration: 'CallSignatureDeclaration',
  ClassImplements: 'ClassImplements',
  ConditionalType: 'ConditionalType',
  ConstructSignatureDeclaration: 'ConstructSignatureDeclaration',
  ConstructorType: 'ConstructorType',
  EmptyBodyFunctionExpression: 'EmptyBodyFunctionExpression',
  ExportAssignment: 'ExportAssignment',

  ExportKeyword: 'ExportKeyword',
  ExternalModuleReference: 'ExternalModuleReference',
  FunctionType: 'TSFunctionType',
  ImportEqualsDeclaration: 'TSImportEqualsDeclaration',
  ImportExpression: 'ImportExpression',
  ImportType: 'ImportType',
  IndexSignature: 'IndexSignature',
  IndexedAccessType: 'IndexedAccessType',
  InferType: 'InferType',

  InterfaceBody: 'InterfaceBody',
  InterfaceHeritage: 'InterfaceHeritage',
  IntersectionType: 'IntersectionType',
  LiteralType: 'TSLiteralType',
  MappedType: 'MappedType',
  MethodSignature: 'MethodSignature',
  ModuleBlock: 'ModuleBlock',
  ModuleDeclaration: 'TSModuleDeclaration',
  NamespaceExportDeclaration: 'NamespaceExportDeclaration',
  NeverKeyword: 'TSNeverKeyword',

  NullKeyword: 'TSNullKeyword',
  NumberKeyword: 'TSNumberKeyword',
  ObjectKeyword: 'ObjectKeyword',
  OptionalType: 'OptionalType',

  ParenthesizedType: 'ParenthesizedType',
  PrivateKeyword: 'PrivateKeyword',
  PropertySignature: 'PropertySignature',
  ProtectedKeyword: 'ProtectedKeyword',
  PublicKeyword: 'PublicKeyword',
  QualifiedName: 'QualifiedName',
  ReadonlyKeyword: 'ReadonlyKeyword',
  RestType: 'RestType',
  StaticKeyword: 'StaticKeyword',

  SymbolKeyword: 'SymbolKeyword',
  ThisType: 'ThisType',
  TupleType: 'TSTupleType',

  JSDocNullableType: 'JSDocNullableType',
  JSDocUnknownType: 'JSDocUnknownType',
  TypeLiteral: 'TypeLiteral',
  TypeOperator: 'TypeOperator',
  TypeParameter: 'TypeParameter',
  TypeParameterDeclaration: 'TypeParameterDeclaration',
  TypeParameterInstantiation: 'TypeParameterInstantiation',
  TypePredicate: 'TypePredicate',
  TypeQuery: 'TypeQuery',
  TypeReference: 'TSTypeReference',
  UnaryExpression: 'UnaryExpression',
  UndefinedKeyword: 'TSUndefinedKeyword',
  UnionType: 'UnionType',
  UnknownKeyword: 'UnknownKeyword',
  UpdateExpression: 'UpdateExpression',
  VariableDeclaration: 'VariableDeclaration',
  VariableDeclarator: 'VariableDeclarator',
  VoidKeyword: 'TSVoidKeyword',
  WhileStatement: 'WhileStatement',
  WithStatement: 'WithStatement',
  YieldExpression: 'YieldExpression',
};
