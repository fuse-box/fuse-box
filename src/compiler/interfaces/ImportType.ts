export enum ImportType {
  REQUIRE = 1,
  FROM = 2,
  RAW_IMPORT = 3,
  DYNAMIC = 4,
}
