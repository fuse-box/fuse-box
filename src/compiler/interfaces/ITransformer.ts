import { ICompilerOptions } from '../../compilerOptions/interfaces';
import { Context } from '../../core/context';
import { IModule } from '../../moduleResolver/module';
import { IProductionContext } from '../../production/ProductionContext';
import { IVisit, IVisitorMod } from '../Visitor/Visitor';
import { ASTNode } from './AST';
import { ImportType } from './ImportType';

export type ITranformerCallback = (visit: IVisit) => undefined | IVisitorMod | void;

export interface ITransformerVisitors {
  onEachNode?: ITranformerCallback;
  onTopLevelTraverse?: ITranformerCallback;
}

export interface IRequireStatementModuleOptions {
  breakDependantsCache?: boolean;
}
export interface ITransformerCommon {
  compilerOptions: ICompilerOptions;
  ctx: Context;
  module: IModule;
  onRequireCallExpression?: (
    importType: ImportType,
    node: ASTNode,
    moduleOptions?: IRequireStatementModuleOptions,
  ) => void;
}

export interface ITransformerProduction {
  ctx: Context;
  module: IModule;
  productionContext: IProductionContext;
}

export interface ITransformer {
  target?: {
    test?: RegExp;
    type?: 'js' | 'js_ts' | 'ts';
  };
  commonVisitors?: (props?: ITransformerCommon) => ITransformerVisitors;
  productionWarmupPhase?: (props?: ITransformerProduction) => ITransformerVisitors;
}
