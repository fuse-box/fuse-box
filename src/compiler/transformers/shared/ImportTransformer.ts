import * as path from 'path';
import { GlobalContext } from '../../program/GlobalContext';
import { ITransformer } from '../../program/transpileModule';
import { createRequireStatement } from '../../Visitor/helpers';
import { IVisit } from '../../Visitor/Visitor';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { ImportType } from '../../interfaces/ImportType';

// FIX SHORTHANDS
export function ImportTransformer(options?: ITransformerSharedOptions): ITransformer {
  options = options || {};
  return {
    onTopLevelTraverse: (visit: IVisit) => {
      const node = visit.node;
      const global = visit.globalContext as GlobalContext;

      if (node.type === 'ImportDeclaration') {
        // converts "./foo/bar.hello.js" to foo_bar_hello_js_1 (1:1 like typescript does)
        const variable = path.basename(node.source.value).replace(/\.|-/, '_') + '_' + global.getNextIndex();

        node.specifiers.forEach(specifier => {
          if (specifier.type === 'ImportSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: specifier.imported.name,
            };
          } else if (specifier.type === 'ImportDefaultSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: 'default',
            };
          } else if (specifier.type === 'ImportNamespaceSpecifier') {
            // only if we have more than one specifier
            // for instance
            // i=mport MySuperClass, * as everything from "module-name";

            // in every other case like:
            // import * as everything from "module-name";
            // we don't need to do anything, since the variables should match

            global.identifierReplacement[specifier.local.name] = {
              first: variable,
            };

            //if (node.specifiers.length > 1) afterStatement = defineVariable(variable, specifier.local.name);
          }
        });

        return {
          onComplete: () => {
            const reqStatement = createRequireStatement(node.source.value, node.specifiers.length && variable);
            // when everything is finished we need to check if those variables have been used at all
            // they were all unused we need remove the require/import statement at all

            // assuming we have no specififers and this import HAS side effects
            // e.g import "./module"
            const parent = visit.parent;
            const property = visit.property;
            if (node.specifiers.length === 0) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.RAW_IMPORT, reqStatement.reqStatement);
                }
                parent[property].splice(index, 1, reqStatement.statement);
              }
              return;
            }

            let atLeastOneInUse = false;
            for (const specifier of node.specifiers) {
              const traced = global.identifierReplacement[specifier.local.name];
              if (traced && traced.inUse) {
                atLeastOneInUse = true;
                break; // we just need to know if we need to keep the node
              }
            }
            // doing a manual replace
            if (atLeastOneInUse) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                parent[property].splice(index, 1, reqStatement.statement);
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.FROM, reqStatement.reqStatement);
                }
              }
            } else {
              const index = parent[property].indexOf(node);
              if (index > -1) parent[property].splice(index, 1);
            }
          },
        };
      }
    },
  };
}
