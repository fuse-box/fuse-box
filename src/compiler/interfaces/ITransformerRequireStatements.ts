import { ASTNode } from './AST';
import { IRequireStatementModuleOptions } from './ITransformer';
import { ImportType } from './ImportType';

export interface ITransformerRequireStatement {
  importType: ImportType;
  moduleOptions?: IRequireStatementModuleOptions;
  statement: ASTNode;
  value?: string;
}
export type ITransformerRequireStatementCollection = Array<ITransformerRequireStatement>;
