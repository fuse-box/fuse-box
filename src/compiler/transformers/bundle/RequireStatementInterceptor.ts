import { BUNDLE_RUNTIME_NAMES } from '../../../bundleRuntime/bundleRuntimeCore';
import { ISchema } from '../../core/nodeSchema';
import { ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { ImportType } from '../../interfaces/ImportType';

export function RequireStatementInterceptor(): ITransformer {
  return {
    commonVisitors: props => {
      return {
        onEach: (schema: ISchema) => {
          const { getLocal, localIdentifier, node, parent, replace } = schema;

          if (localIdentifier) {
            if (node.name === 'require' && parent.type !== ASTType.CallExpression) {
              if (getLocal(node.name)) return;
              return replace({ name: BUNDLE_RUNTIME_NAMES.ARG_REQUIRE_FUNCTION, type: 'Identifier' });
            }
          }

          // handle typeof
          if (node.operator === 'typeof' && node.type === ASTType.UnaryExpression) {
            if (node.argument && node.argument.name) {
              const name = node.argument.name;
              // we must preserve local variable
              if (name === 'require') {
                if (getLocal(name)) return;
                return replace({ type: 'Literal', value: 'function' });
              }
            }
          }

          if (!props.onRequireCallExpression) return;

          if (node.type === 'CallExpression' && node.callee.name === 'require' && !node['emitted']) {
            if (!getLocal('require')) props.onRequireCallExpression(ImportType.REQUIRE, node);
          }
          return;
        },
      };
    },
  };
}
