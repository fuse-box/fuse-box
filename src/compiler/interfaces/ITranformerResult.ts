import { ASTNode } from './AST';
import { ITransformerRequireStatementCollection } from './ITransformerRequireStatements';

export interface ITransformerResult {
  ast?: ASTNode;
  requireStatementCollection: ITransformerRequireStatementCollection;
}
