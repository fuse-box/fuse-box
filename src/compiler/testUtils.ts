import { ICompilerOptions } from '../compilerOptions/interfaces';
import { transformModule } from './core/transformModule';
import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ITransformer, ITransformerCommon } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { ImportType } from './interfaces/ImportType';
import { parseTypeScript } from './parser';
import { ISerializableTransformationContext } from './transformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  code: string;
  compilerOptions?: ICompilerOptions;
  jsx?: boolean;
  props?: ISerializableTransformationContext;
  transformers: Array<ITransformer>;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });
  const userProps: ISerializableTransformationContext = props.props || {};
  userProps.compilerOptions = props.compilerOptions || {};
  const visitorProps: ITransformerCommon = { onRequireCallExpression, transformationContext: userProps };

  const tranformers = [];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors(visitorProps));
    }
  }

  transformModule({ root: ast, transformers: tranformers });

  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
