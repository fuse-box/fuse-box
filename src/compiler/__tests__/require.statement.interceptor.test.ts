import { ImportType } from '../interfaces/ImportType';
import { initCommonTransform } from '../testUtils';
import { RequireStatementInterceptor } from '../transformers/bundle/RequireStatementInterceptor';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';

const testTranspile = (props: { code: string; fileName?: string; target?: string }) => {
  return initCommonTransform({
    code: props.code,
    transformers: [RequireStatementInterceptor(), ImportTransformer()],
  });
};

describe('Require statement intercepto', () => {
  it('should emit require', () => {
    const result = testTranspile({
      code: `
      console.log(require("./a"))
    `,
    });

    expect(result.requireStatementCollection).toEqual([
      {
        importType: ImportType.REQUIRE,
        statement: {
          arguments: [{ type: 'Literal', value: './a' }],
          callee: { name: 'require', type: 'Identifier' },
          optional: false,
          type: 'CallExpression',
        },
      },
    ]);
  });

  it('should emit only twice', () => {
    const result = testTranspile({
      code: `
      import hey from "./hey";
      console.log(require("./a"), hey)
    `,
    });
    expect(result.requireStatementCollection).toEqual([
      {
        importType: ImportType.REQUIRE,
        statement: {
          optional: false,
          type: 'CallExpression',

          arguments: [{ type: 'Literal', value: './a' }],
          callee: { name: 'require', type: 'Identifier' },
        },
      },
      {
        importType: ImportType.FROM,
        statement: {
          arguments: [{ type: 'Literal', value: './hey' }],
          callee: { name: 'require', type: 'Identifier' },
          type: 'CallExpression',
        },
      },
    ]);
  });

  it('it should honor custom require', () => {
    const result = testTranspile({
      code: `
      function foo(require) {
        const b = require("./a");
        console.log(b);
      }
      console.log(foo);
    `,
    });
    expect(result.requireStatementCollection).toHaveLength(0);
  });

  it('it should replace typeof require', () => {
    const result = testTranspile({
      code: `
      console.log(typeof require);
    `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('it should replace typeof require is defined in the scope', () => {
    const result = testTranspile({
      code: `
      function some(require){
        console.log(typeof require);
      }
    `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('it should replace require ref', () => {
    const result = testTranspile({
      code: `
      if (typeof require === 'function' && require) {
        console.log(1);
      }
    `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('it should replace require ref 2', () => {
    const result = testTranspile({
      code: `
      if (typeof require === 'function' && typeof require.ensure) {
        console.log(1);
      }
    `,
    });

    expect(result.code).toMatchSnapshot();
  });
});
