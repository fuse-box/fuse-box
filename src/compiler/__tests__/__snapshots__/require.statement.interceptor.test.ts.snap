// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Require statement intercepto it should replace require ref 1`] = `
"if (\\"function\\" === 'function' && __fusereq) {
  console.log(1);
}
"
`;

exports[`Require statement intercepto it should replace require ref 2 1`] = `
"if (\\"function\\" === 'function' && typeof __fusereq.ensure) {
  console.log(1);
}
"
`;

exports[`Require statement intercepto it should replace typeof require 1`] = `
"console.log(\\"function\\");
"
`;

exports[`Require statement intercepto it should replace typeof require is defined in the scope 1`] = `
"function some(require) {
  console.log(typeof require);
}
"
`;
